### ISSUE TITLE
Improving `basenc` / `base32` / `base64`

### ISSUE THREAD

**Issue by bal-e** [uutils/coreutils#5698]
I've identified a few issues with the encoding/decoding tools that I'm interested in solving:

1. They are not iterative / streaming.  The entirety of the input is read first before it is all decoded.  One the one hand, this allows you to detect errors upfront before you've output anything.  But it means that these tools can't be used for iterative computations (e.g. over `while` loops) and they may consume large amounts of memory.  The GNU implementations of these tools are streaming as well, so I'd classify this as an incompatibility bug.

2. Significant optimization potential is available.  I have a fair bit of experience with SIMD and I've designed AVX2 and AVX512 algorithms for Base32 / Base64 encoding - it's difficult but I would like to see if there are performance benefits to it, particularly for large inputs.  Optimizations can be added iteratively, and can be feature-gated or even based on runtime CPU feature detection.

3. Compiled binaries for `base32` / `base64`, on my machine, currently take up 23 MiB in debug mode and 3 MiB in release mode.  This is surprisingly large and I believe the majority of that work lies in the argument parsing rather than the actual encoding.  Since `basenc` / `base32` / `base64` have essentially the same arguments to deal with, they can be combined.

To this end, here's my proposal:

1. Rewrite `src/uucore/src/lib/features/encoding.rs` to expose a streaming rather than one-shot interface.  Along the way, I can add a sub-module for each encoding format so that per-format optimization routines can be defined.  Runtime CPU detection, if implemented, would be format-generic.

2. Setup `basenc` as a multicall binary so that `base32` / `base64` are symlinks to it.  The argument parsing for `basenc` will have to be updated to only expose format-selecting arguments if `argv[0]` is `basenc`.

3. Implement format-specific optimization routines over time, supported by benchmarking.  For example, if the user only provides less than 1 KiB of data, a less optimized routine can be used (advanced AVX512 instructions have some interesting effects on CPU clock frequencies, so we should avoid them unless a large amount of input needs to be processed).

It may be preferable that I write a new, independent crate for optimized data encoding and create a dependency on that (rather than the current `data-encoding` crate); the alternative is to eliminate the external dependency completely.

I'd like to ask for prior approval from the `uutils` community before I start working on this.  Note that task 1 and 2 are effectively independent - while I'm more interested in the first, the second seems more important.  Maintainers, what do you think?

**Comment by tertsdiepraam** [comment]
If the encoding is in a separate crate then I thjnk they should already share a a lot of code and then then the multicall might not be necessary. The only difference would be the construction of the clap command. I think I'd rather keep them as separate crates but with a shared crate underneath.

If we set up that structure first, then we can make that shared crate streaming, because that is indeed a bug. Performance optimizations are welcome, but in my opinion low priority and should not increase the complexity too much. The unoptimized baseline implementation is the most important. So, I'd propose to make it streaming without SIMD etc. first.

**Comment by bal-e** [comment]
> If the encoding is in a separate crate then I thjnk they should already share a a lot of code and then then the multicall might not be necessary.

They already share a lot in terms of code, I agree - but due to static linking, the same code is duplicated in every executable.  A multicall binary would eliminate that overhead.

> Performance optimizations are welcome, but in my opinion low priority and should not increase the complexity too much.

My first priority is the streaming interface, I agree.  But I'd like to provide some more complex implementations over time - while providing simple fallbacks, of course.  If you don't want that complexity in this repository, the shared crate solution seems best.

**Comment by tertsdiepraam** [comment]
Yes a multicall binary would help, but that's what we have the general multicall binary for already. I think multiple levels of multicall binaries is a bit too much. We have this for the checksum utils and I've been wanting go remove that for a while 😄

**Comment by bal-e** [comment]
Alright, then I'll focus on the streaming interface for now.  Hopefully better argument parsing systems can bring the per-binary overhead down.

**Comment by bal-e** [comment]
Status update: I've got basic implementations of all the formats together in my crate [`glassy`](https://git.sr.ht/~bal-e/glassy) ([documentation](https://docs.rs/glassy)).  I've got a pretty thorough automated testing setup, but I need to test some failure cases better.  Once that's done, I'll polish up the documentation a bit and begin working on a PR integrating it here.

**Comment by sylvestre** [comment]
@bal-e i am sorry but I am usually quite reluctant to add a new dependency on brand new crate :(
I would prefer you integrate this code directly into the coreutils

**Comment by bal-e** [comment]
@sylvestre There are a few reasons why I think this should be a separate dependency:

1. The code being developed forms a good utility library for any Rust project that needs to perform Base32 encoding/decoding.  I already have a fairly stable API for `glassy`, and it will not have any dependencies of its own, so I don't believe that using `glassy` will incur any significant maintenance burden on `uutils`.

2. Over time, I will be developing `glassy` further to make it much more performant (this is possible because it only offers the exact formats needed for `basenc` - it is not overly flexible).  `uutils` should be able to benefit from these performance improvements without having to maintain the code itself.

3. With this change, I will be consolidating two existing dependencies of `uutils` (`data-encoding` and `z85`), so this will actually reduce `uutils`' total number of dependencies.

4. Should `uutils` opt to implement new conversion formats, they can be easily integrated into `glassy`.  `glassy` operates through `Encode` / `Decode` traits, which anybody can implement, and it exposes all of the utilities its own implementations urkise for convenience, so new formats can be implemented easily both inside and outside `glassy` (w.  I am happy to introduce support for new formats into `glassy` or to make the existing formats more flexible.

As it stands, I don't believe that `glassy`'s position as a "brand new crate" should affect how reliable it is.  As a standalone Base32/etc. converter, the standards it implements are concrete and unchanging.  Even if I were to stop maintaining it, the actual code within will continue working for the foreseeable future.  At that stage, it may make sense to fork it and maintain a copy within `uutils`, but I do plan on maintaining it thoroughly.

**Comment by sylvestre** [comment]
Yeah, we could move it under the uutils umbrella. Otherwise, I am quite reluctant to add a dependency like this despite all its qualities. 
I am here to maintain rust-coreutils for the years to come and I am careful when adding new dependencies.

**Comment by bal-e** [comment]
I understand that moving `glassy` under the uutils organization will make it easier for you to maintain it in the future.  But I would appreciate it if you could use the crate as published from my repository.  If I ever stop maintaining it, feel free to fork it here.

P.S. The `z85` crate that `uutils` currently depends on was last updated on April 12, 2022.  Using the crate I have published may not be perfect from a maintenance perspective, but it should certainly be a better choice than relying on 2-year old code.

**Comment by andrewliebenow** [comment]
No SIMD, but performance is roughly equal to or superior to GNU Core Utilities with these changes:

https://github.com/uutils/coreutils/pull/6719

The main point of weakness is line wrapped encoding (still slower than GNU Core Utilities).

**Comment by romanstingler** [comment]
now that SIMD was introduced I wanted to have a look at the current situation 
in `src/uu/basenc/BENCHMARKING.md`

there is a section 

➖ Slower than GNU Core Utilities

```
❯ poop \
    '/usr/bin/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte' \
    './target/release/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte'

Benchmark 1 (6 runs): /usr/bin/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           836ms ± 13.3ms     822ms …  855ms          0 ( 0%)        0%
  peak_rss           2.05MB ± 73.0KB    1.94MB … 2.12MB          0 ( 0%)        0%
  cpu_cycles         2.85G  ± 32.8M     2.82G  … 2.91G           0 ( 0%)        0%
  instructions       14.0G  ± 58.7      14.0G  … 14.0G           0 ( 0%)        0%
  cache_references   70.0M  ± 6.48M     63.7M  … 78.8M           0 ( 0%)        0%
  cache_misses        582K  ±  172K      354K  …  771K           0 ( 0%)        0%
  branch_misses       667K  ± 4.55K      662K  …  674K           0 ( 0%)        0%
Benchmark 2 (6 runs): ./target/release/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           884ms ± 6.38ms     878ms …  895ms          0 ( 0%)        💩+  5.7% ±  1.6%
  peak_rss           2.65MB ± 66.8KB    2.55MB … 2.74MB          0 ( 0%)        💩+ 29.3% ±  4.4%
  cpu_cycles         3.15G  ± 8.61M     3.14G  … 3.16G           0 ( 0%)        💩+ 10.6% ±  1.1%
  instructions       10.5G  ±  275      10.5G  … 10.5G           0 ( 0%)        ⚡- 24.9% ±  0.0%
  cache_references   93.5M  ± 6.10M     87.2M  …  104M           0 ( 0%)        💩+ 33.7% ± 11.6%
  cache_misses        415K  ± 52.3K      363K  …  474K           0 ( 0%)          - 28.8% ± 28.0%
  branch_misses      1.43M  ± 4.82K     1.42M  … 1.43M           0 ( 0%)        💩+113.9% ±  0.9%
```

I ran that on my machine and it seems that we are reading the whole 1GB into mem
`peak_rss 1.08GB `. this might be an issue on embedded devices or containers

```
❯ cd /srv/opensource/coreutils && poop \
                '/usr/bin/basenc --base64 -- /tmp/one-random-gibibyte' \
                'uu-basenc --base64 -- /tmp/one-random-gibibyte'
Benchmark 1 (9 runs): /usr/bin/basenc --base64 -- /tmp/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           583ms ± 18.1ms     565ms …  627ms          1 (11%)        0%
  peak_rss           3.95MB ± 72.5KB    3.84MB … 4.00MB          0 ( 0%)        0%
  cpu_cycles         2.20G  ± 11.6M     2.19G  … 2.23G           1 (11%)        0%
  instructions       13.7G  ± 44.2      13.7G  … 13.7G           0 ( 0%)        0%
  cache_references   94.8M  ±  377K     94.2M  … 95.5M           0 ( 0%)        0%
  cache_misses       5.53M  ± 32.9K     5.47M  … 5.58M           0 ( 0%)        0%
  branch_misses      1.47M  ± 8.39K     1.45M  … 1.48M           0 ( 0%)        0%
Benchmark 2 (12 runs): uu-basenc --base64 -- /tmp/one-random-gibibyte
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           427ms ± 4.30ms     420ms …  434ms          0 ( 0%)        ⚡- 26.7% ±  1.9%
  peak_rss           1.08GB ±  133KB    1.08GB … 1.08GB          0 ( 0%)        💩+27247.1% ±  2.6%
  cpu_cycles         1.10G  ± 10.3M     1.07G  … 1.11G           1 ( 8%)        ⚡- 50.1% ±  0.5%
  instructions       1.91G  ±  796      1.91G  … 1.91G           1 ( 8%)        ⚡- 86.1% ±  0.0%
  cache_references    158M  ±  295K      158M  …  159M           0 ( 0%)        💩+ 66.7% ±  0.3%
  cache_misses       9.63M  ±  109K     9.43M  … 9.80M           0 ( 0%)        💩+ 74.1% ±  1.4%
  branch_misses       339K  ± 16.8K      314K  …  372K           0 ( 0%)        ⚡- 76.9% ±  0.9%
```

**Comment by sylvestre** [comment]
could you please add benchmark (in a sep PR) ?

see  src/uu/*/benches for example

**Comment by Nugine** [comment]
`base64-simd` author here. CC me if you need any help!
